changes.go

```diff
@@ -22,54 +22,294 @@ func FromData(data *charm.BundleData) []*Change {
 }
 
 // Change holds a single change required to deploy a bundle.
-type Change struct {
-	// Id is the unique identifier for this change.
-	Id string `json:"id"`
-	// Method is the action to be performed to apply this change.
-	Method string `json:"method"`
-	// Args holds a list of arguments to pass to the method.
-	Args []interface{} `json:"args"`
-	// Requires holds a list of dependencies for this change. Each dependency
+type Change interface {
+	// Id returns the unique identifier for this change.
+	Id() string
+	// Requires returns a list of dependencies for this change. Each dependency
 	// is represented by the corresponding change id, and must be applied
 	// before this change is applied.
-	Requires []string `json:"requires"`
+	Requires() []string
+	// Method returns the action to be performed to apply this change.
+	Method() string
+	// GUIArgs returns positional arguments to pass to the method, suitable for
+	// being serialized and sent to the Juju GUI.
+	GUIArgs() []interface{}
+	// setId is used to set the identifier for the change.
+	setId(string)
 }
 
-// changeset holds the list of changes returned by FromData.
-type changeset struct {
-	changes []*Change
+type changeInfo struct {
+	id       string
+	requires []string
+	method   string
+}
+
+// Id implements Change.Id.
+func (ch *changeInfo) Id() string {
+	return ch.id
+}
+
+// Requires implements Change.Requires.
+func (ch *changeInfo) Requires() []string {
+	if ch.requires == nil {
+		return []string{}
+	}
+	return ch.requires
+}
+
+// Method implements Change.Method.
+func (ch *changeInfo) Method() string {
+	return ch.method
+}
+
+// setId implements Change.setId.
+func (ch *changeInfo) setId(id string) {
+	ch.id = id
+}
+
+// newAddCharmChange creates a new change for adding a charm.
+func newAddCharmChange(args AddCharmArgs, requires ...string) *AddCharmChange {
+	return &AddCharmChange{
+		changeInfo: changeInfo{
+			requires: requires,
+			method:   "addCharm",
+		},
+		Args: args,
+	}
+}
+
+// AddCharmChange holds a change for adding a charm to the environment.
+type AddCharmChange struct {
+	changeInfo
+	// Args holds parameters for adding a charm.
+	Args AddCharmArgs
+}
+
+// GUIArgs implements Change.GUIArgs.
+func (ch *AddCharmChange) GUIArgs() []interface{} {
+	return []interface{}{ch.Args.Charm}
+}
+
+// AddCharmArgs holds parameters for adding a charm to the environment.
+type AddCharmArgs struct {
+	// Charm holds the URL of the charm to be added.
+	Charm string
+}
+
+// newAddMachineChange creates a new change for adding a machine or container.
+func newAddMachineChange(args AddMachineArgs, requires ...string) *AddMachineChange {
```

Comment:
One thought: instead of having requires as an argument to every change constructor,
we could implement setRequires on changeInfo and call it after the constructor is called.